File: jar/src/main/java/jline/api/mc/Smp_passage_lst.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */
package jline.api.mc;

import java.util.List;
import java.util.function.UnaryOperator;

import org.apache.commons.math3.complex.Complex;

import jline.util.matrix.Matrix;

/**
 * Laplace-Stieltjes transform of the first passage time into a target state set
 * for a SEMI-MARKOV chain.
 *
 * <p>Reference: P. G. Harrison and W. J. Knottenbelt, "Passage Time
 * Distributions in Large Markov Chains", 2002, Eqs. 4-5:
 *
 * <pre>
 *     L_i(s) = sum_{k not in B} r*_ik(s) L_k(s) + sum_{k in B} r*_ik(s),
 * </pre>
 *
 * so (I - R*_AA(s)) L_A(s) = R*_AB(s) 1, one linear system per value of s. The
 * sojourn transforms are given per state, so r*_ik(s) = P(i,k) h*_i(s) and the
 * complex numbers stay on the DIAGONAL of the system, which is the easier of
 * the two cases the paper distinguishes.
 *
 * <p>Distribution objects supply their own transform: {@code Markovian.evalLST}
 * gives the closed form pie (sI-D0)^-1 (-D0) e for the phase-type family, so
 * {@code s -> new Complex(dist.evalLST(s.getReal()), 0)} is NOT enough for a
 * complex argument -- pass a genuinely complex transform.
 */
public final class Smp_passage_lst {

    private Smp_passage_lst() {
    }

    /**
     * @param hlst one sojourn transform h*_i(s) per state, complex-argument
     */
    public static Complex smp_passage_lst(Matrix P, List<UnaryOperator<Complex>> hlst, Matrix pi0,
                                          int[] target, Complex s) {
        int n = P.getNumRows();
        int[] tgt = PassageSupport.uniqueTarget(target, n, "smp_passage_lst");
        if (hlst == null || hlst.size() != n) {
            throw new RuntimeException(
                    "smp_passage_lst: hlst must carry one transform per state");
        }
        int[] A = PassageSupport.complement(tgt, n);
        int nA = A.length;

        double atom = 0.0;
        boolean weighted = pi0 != null && pi0.length() == n;
        if (weighted) {
            for (int k : tgt) {
                atom += pi0.get(k);
            }
        }

        Complex[] h = new Complex[nA];
        for (int a = 0; a < nA; a++) {
            h[a] = hlst.get(A[a]).apply(s);
        }
        Complex[][] M = new Complex[nA][nA];
        Complex[] b = new Complex[nA];
        for (int a = 0; a < nA; a++) {
            for (int c = 0; c < nA; c++) {
                Complex diag = new Complex((a == c) ? 1.0 : 0.0, 0.0);
                M[a][c] = diag.subtract(h[a].multiply(P.get(A[a], A[c])));
            }
            double pb = 0.0;
            for (int k : tgt) {
                pb += P.get(A[a], k);
            }
            b[a] = h[a].multiply(pb);
        }
        Complex[] x = PassageSupport.solveComplex(M, b, "smp_passage_lst");
        Complex acc = new Complex(atom, 0.0);
        for (int a = 0; a < nA; a++) {
            double w = weighted ? pi0.get(A[a]) : 1.0 / n;
            acc = acc.add(x[a].multiply(w));
        }
        return acc;
    }
}

File: matlab/src/api/mc/smp_passage_lst.m
function L = smp_passage_lst(P, hlst, pi0, target, s)
% L = SMP_PASSAGE_LST(P, HLST, PI0, TARGET, S)
%
% Laplace-Stieltjes transform of the first passage time into the target state
% set for a semi-Markov chain, at the (possibly complex) points S.
%
% Reference: P. G. Harrison and W. J. Knottenbelt, "Passage Time Distributions
% in Large Markov Chains", 2002, Eqs. 4-5:
%
%     L_i(s) = sum_{k not in B} r*_ik(s) L_k(s) + sum_{k in B} r*_ik(s)
%
% so (I - R*_AA(s)) L_A(s) = R*_AB(s) 1, one linear system per value of s.
%
% HLST is either
%   (nstates x 1) cell of handles h*_i(s)   the sojourn in i depends only on i,
%                                           so r*_ik(s) = P(i,k) h*_i(s) and
%                                           the complex numbers stay on the
%                                           DIAGONAL of the system (Eq. 5)
%   (nstates x nstates) cell of r*_ik(s)    the full Markov-renewal kernel; the
%                                           coefficients are then complex
%                                           functions of s throughout, which is
%                                           the harder case the paper flags
%
% Distribution objects supply their own transform: MARKOVIAN.evalLST gives the
% closed form pie (sI-D0)^{-1} (-D0) e for the phase-type family, so
% hlst{i} = @(s) dist.evalLST(s) is the intended way to build these.
%
% Copyright (c) 2012-2026, Imperial College London
% All rights reserved.

n = size(P,1);
target = unique(reshape(target,1,[]));
isTarget = false(1,n);
isTarget(target) = true;
A = find(~isTarget);
nA = numel(A);

if isempty(pi0)
    pi0 = ones(1,n)/n;
end
pi0 = reshape(pi0,1,[]);
if numel(pi0) ~= n
    line_error(mfilename, 'PI0 must be a distribution over the state space, one entry per state.');
end
atom = sum(pi0(target));

perState = iscell(hlst) && (size(hlst,2) == 1 || isvector(hlst));
L = zeros(size(s));
I = eye(nA);
for is = 1:numel(s)
    sv = s(is);
    if perState
        h = zeros(nA,1);
        for a = 1:nA
            h(a) = hlst{A(a)}(sv);
        end
        RAA = diag(h) * P(A,A);
        RAB = diag(h) * sum(P(A,target),2);
    else
        Rs = zeros(n,n);
        for i = 1:n
            for k = 1:n
                if ~isempty(hlst{i,k})
                    Rs(i,k) = hlst{i,k}(sv);
                end
            end
        end
        RAA = Rs(A,A);
        RAB = sum(Rs(A,target),2);
    end
    LA = (I - RAA) \ RAB;
    L(is) = pi0(A) * LA + atom;
end
end
